lib/rules/id-match.js

```diff
@@ -238,6 +238,17 @@ module.exports = {
                         }
                     }
 
+                    // For https://github.com/eslint/eslint/issues/15123
+                    if (parent.parent && parent.parent.type === "ObjectExpression") {
+                        if (
+                            (checkProperties && onlyDeclarations && !parent.computed) &&
+                            parent.key === node &&
+                            isInvalid(name)
+                        ) {
+                            report(node);
+                        }
+                    }
```

Comment:
This looks confusing, why should `onlyDeclarations` affect whether or not identifiers used as property names in object literals should be reported?

Reply:
I think it would be best to extract handling of property names in object literals into a separate `else-if` in the main `if-else-if` chain. The logic in `if (parent.type === "Property" || parent.type === "AssignmentPattern")` is already nearly impossible to follow as it handles too many cases that have similar AST representation but are logically completely different.

```js
}  else if (
        parent.type === "Property" &&
        parent.parent.type === "ObjectExpression" &&
        parent.key === node &&
        !parent.computed
    ) {
    if (checkProperties && isInvalid(name)) {
        report(node);
    }
} else if (parent.type === "Property" || parent.type === "AssignmentPattern") {
    // ...
}
```

Reply:
I have refactored the code 👍🏻